src/routes/oauth/index.ts

```diff
@@ -0,0 +1,296 @@
+import { v4 as uuidv4 } from 'uuid';
+import express, { Router } from 'express';
+import session from 'express-session';
+import grant from 'grant';
+
+import { ERRORS, sendError } from '@/errors';
+import {
+  ENV,
+  generateRedirectUrl,
+  getNewRefreshToken,
+  getUserByEmail,
+  gqlSdk,
+  insertUser,
+} from '@/utils';
+import {
+  queryValidator,
+  redirectTo as redirectToRule,
+  registrationOptions,
+} from '@/validation';
+import { SessionStore } from './session-store';
+import { logger } from '@/logger';
+import { InsertUserMutation } from '@/utils/__generated__/graphql-request';
+import {
+  createGrantConfig,
+  normaliseProfile,
+  preRequestProviderMiddleware,
+  transformOauthProfile,
+} from './utils';
+import { OAUTH_ROUTE } from './config';
+
+const SESSION_NAME = 'connect.sid';
+
+/**
+ * We create a Grant configuration when the service starts,
+ * so it determines which providers are enabled.
+ * If a provider is enabled but incorrectly configured, the service will start,
+ * but the end user will be redirected with an error to the client app.
+ */
+const grantConfig = createGrantConfig();
+
+/**
+ * GET /signin/provider/{provider}
+ * @summary
+ * @param {string} provider.path.required - name param description - enum:github,google,facebook,twitter,apple,azuread,windowslive,linkedin,spotify,strava,gitlab,bitbucket,twitch
+ * @param {string} redirectUrl.query.required -
+ * @return {string} 302 - Redirect to the provider's authentication page
+ * @tags Authentication
+ */
+
+/**
+ * GET /signin/provider/{provider}/callback
+ * @summary Oauth callback url, that will be used by the Oauth provider, to redirect to the client application. Attention: all providers are using a GET operation, except Apple and Azure AD that use POST
+ * @param {string} provider.path.required - name param description - enum:github,google,facebook,twitter,apple,azuread,windowslive,linkedin,spotify,strava,gitlab,bitbucket,twitch
+ * @param {string} redirectUrl.query.required
+ * @return {string} 302 - Redirect to the initial url given as a query parameter in /signin/provider/{provider}
+ * @tags Authentication
+ */
+export const oauthProviders = Router()
+  /**
+   * Use a middleware to keep the session between Oauth requests.
+   * Once the authentication choregraphy is done (either success or failure), the session is destroyed.
+   */
+  .use(
+    session({
+      secret: 'grant',
+      resave: false,
+      saveUninitialized: true,
+      store: new SessionStore(),
```

Comment:
Do we need to use `express-session` here? Just wondering if we could do without that package. The extension of the Store seems pretty trivial.

Reply:
Yes, we still need it so it handles the session during the oauth choreography